src/network/providers/provider.ts

```diff
@@ -0,0 +1,141 @@
+    return Object.values(ChainId).includes(chainId);
+}
+
+/**
+ * BlockchainProvider is the abstraction that consolidates and exposes
+ * blockchain functionality and provides a unified solution for working
+ * with different chains such as Ethereum and Solana.
+ */
+class BlockchainProvider implements Provider {
+    private providers: {[key in Blockchain]?: Provider} = {};
+
+    private getProvider(chainId: ChainId): Provider {
+        const blockchain = chainToBlockchain[chainId];
+        if (!this.providers[blockchain]) {
+            switch (blockchain) {
+                case Blockchain.ETHEREUM:
+                    this.providers[blockchain] = EthereumProvider.getInstance();
+                    break;
+                case Blockchain.SOLANA:
+                    this.providers[blockchain] = SolanaProvider.getInstance();
+                    break;
+                default:
+                    const errMsg = `Unsupported chain ID "${chainId}".`;
+                    log.error({chainId}, errMsg);
+                    throw new Error(errMsg);
+            }
+        }
+        return this.providers[blockchain] as Provider;
+    }
+
+    private notImplementedErr(chainId: ChainId, method: string): Error {
+        const errMsg = `Method ${method} not implemented by ${this.name(chainId)}`;
+        log.error({chainId, method}, errMsg);
+        return new Error(errMsg);
+    }
+
+    public name(chainId: ChainId): string {
+        return this.getProvider(chainId).name(chainId);
+    }
+
+    public async resolveDomain(chainId: ChainId, domain: string): Promise<DomainRegistry> {
+        const registry = await this.getProvider(chainId).resolveDomain(chainId, domain);
+        return registry;
+    }
+
+    public async setDomainContent(
+        chainId: ChainId,
+        domain: string,
+        data: string
+    ): Promise<Record<string, string|number|boolean>> {
+        const tx = await this.getProvider(chainId).setDomainContent(chainId, domain, data);
+        return tx;
+    }
+
+    public async send(chainId: ChainId, data: RPCRequest): Promise<JsonRpcResponse> {
+        const provider = this.getProvider(chainId);
+        const resp = await provider.send(chainId, data);
+        return resp;
+    }
+
+    public async sendTransaction(
+        chainId: ChainId,
+        reqId: number,
+        txParams: unknown[]
+    ): Promise<JsonRpcResponse> {
+        const resp = await this.getProvider(chainId).sendTransaction(chainId, reqId, txParams);
+        return resp;
+    }
+
+    public async requestAccounts(chainId: ChainId, reqId: number): Promise<JsonRpcResponse> {
+        const accts = await this.getProvider(chainId).requestAccounts(chainId, reqId);
+        return accts;
+    }
+
+    public async getBlockNumber(chainId: ChainId): Promise<number> {
+        const provider = this.getProvider(chainId);
+        if (!provider.getBlockNumber) {
+            throw this.notImplementedErr(chainId, 'getBlockNumber');
```

Comment:
Since `getBlockNumber` is optional, meaning not all Providers will implement it. We check here and throw a clear error.